Seed fixing

miquelflorensa opened the pull request:
## Description

This PR introduces seed fixing for parameter initialization across all layers. Each layer now includes an additional parameter, seed, which allows for setting a fixed seed for initializing its weights and biases. By default, seed is set to -1, which enables random initialization. If seed is set to a value greater than or equal to 0, it uses that value as the fixed seed.

I also modified the `regression_heteros` to show how to fix the seed.
```
net = Sequential(
        # Adding seed to every layer to make the results reproducible
        Linear(1, 128, seed=1),
        ReLU(),
        Linear(128, 128, seed=1),
        ReLU(),
        Linear(128, 2, seed=1),
        EvenExp()
    )
```

miquelflorensa commented:
@davidwardan @van-dai-vuong Could you please verify that the seed fixing works in your cases? Also, could you review the code to ensure all cases are covered?

lhnguyen102 commented:
@miquelflorensa I just had a discussion with @jamesgoulet about the random seed issue. We come up with a potential solution to address it. We was thinking of using a SeedManager class like below. Note here that calling `SeedManager::getInstance() `multiple times does not create multiple instances but rather accesses the same instance each time. It is just an idea but you can test it out to see if it works. Feel free to make it better.
```cpp
class SeedManager {
public:
    
    static SeedManager& get_instance() {
        static SeedManager instance;
        return instance;
    }

    void manual_seed(unsigned int seed) {
        global_random_engine.seed(seed);
    }

    std::mt19937& get_engine() {
        return global_random_engine;
    }

private:
    SeedManager() : global_random_engine(std::random_device{}()) {}
    // Delete copy constructor and assignment operator
    SeedManager(const SeedManager&) = delete;
    SeedManager& operator=(const SeedManager&) = delete;
    std::mt19937 global_random_engine;  
};

std::tuple<std::vector<float>, std::vector<float>> gaussian_param_init_ni(
    float scale, float gain, float noise_gain, int N)
/* Parameter initialization of TAGI neural network including the noise's hidden
 * states
 *
 * Args:
 *    scale: Standard deviation for weight distribution
 *    gain: Multiplication factor
 *    N: Number of parameters
 *
 * Returns:
 *    m: Mean
 *    S: Variance
 */
{
    // Get generator
    std::mt19937& gen = SeedManager::getInstance().getEngine();

    // Initialize pointers
    std::vector<float> S(N);
    std::vector<float> m(N);

    // Weights
    for (int i = 0; i < N; i++) {
        // Variance for output and noise's hidden states
        if (i < N / 2) {
            S[i] = gain * pow(scale, 2);
        } else {
            S[i] = noise_gain * pow(scale, 2);
            scale = pow(S[i], 0.5);
        }

        // Get normal distribution
        std::normal_distribution<float> d(0.0f, scale);

        // Get sample for weights
        m[i] = d(gen);
    }

    return {m, S};
}
```

lhnguyen102 commented:
> @davidwardan @van-dai-vuong Could you please verify that the seed fixing works in your cases? Also, could you review the code to ensure all cases are covered?

@miquelflorensa I am working on unittests for both Python and C++ (ETA: end of the week), so that we dont need to do manual checks anymore. I'll add you guys as reviewers when it is ready

miquelflorensa commented:
@lhnguyen102 I just added a setter for the seed in the Sequential class that sets the seed for each layer. It works, but it might be a bit messy 😅. Creating a seed manager class could be a good solution too, but we’d lose the option to define a seed for individual layers easily (though I’m not sure if that’s something we’d ever want). Anyway, I’ll work on your solution tomorrow, and we can see which approach might be best.

van-dai-vuong commented:
@miquelflorensa after fixing the seed, I sill have different results for the test results, even if I did not shuffle data.

miquelflorensa commented:
> @miquelflorensa I'll let you to decide this issue. I think that we just want to (1) reproduce the same initial values for weights and biases for each layer given a random seed and (2) generate random and different values for each layer. I really like this [approach](https://pytorch.org/docs/stable/generated/torch.manual_seed.html) from pytorch that sets a random seed for every thing
> 
> ```python
> torch.manual_seed(42)
> ```

@lhnguyen102 Right now I am doing a setter in the sequential class like this:
```
net.manual_seed(0)
```
The problem is that it only works on GPU cause, because it needs to be set before creating the layers. Should I maybe create  a way to set the seed before even creating the model like:
```
pyTAGI_seed(0)
```

lhnguyen102 commented:
> The problem is that it only works on GPU cause, because it needs to be set before creating the layers

@miquelflorensa I don't understand what you meant here. Why it does not work for CPU but GPU?. The way I look at this problem is that you currently  have to modified 45 files in order to achieve this randomization of parameters... a lot of changes for this problem and somehow complicated. I would test the `SeedManager` approach to see if it works because it requires changing a single file `param_init.cpp`

miquelflorensa commented:
@lhnguyen102 Yes, I am currently working on the binding for just a `Seed Manager` class. I will tell you whenever I upload the changes.

miquelflorensa commented:
@lhnguyen102, I’ve added the changes, but I have two questions: should we call it `tagi.manual_seed(0)` instead of `SeedManager.manual_seed(0)`? And for preprocessing functions that involve randomization, should they depend on this manual_seed function, or should we keep them separate and set the seed with `np.random.seed(0)`?

lhnguyen102 commented:
@miquelflorensa  I would call it `pytagi.manual_seed(42)` for python version and for c++ is up to you. In the python side, we will have to have to use the randomization function of external library since we wont provide the preprocessing in pytagi.

lhnguyen102 commented:
The real test will be the scenario where given a random seed, we have the same weights and biases for as many re-run on both CUDA and CPU.

miquelflorensa commented:
> @miquelflorensa I would call it `pytagi.manual_seed(42)` for python version and for c++ is up to you. In the python side, we will have to have to use the randomization function of external library since we wont provide the preprocessing in pytagi.

Done

miquelflorensa commented:
What would you do? Would you initialize the layers during the first forward pass instead of during model construction?

lhnguyen102 commented:
@miquel-florensa As I understand, this  class `SeedManager` is a singleton pattern meanings it lives through out the application running. The random engine will automatically transition to another state when it gets called from any function. I think we can do much more simpler than the current implementation. I would implement  the following function that can serve as  binding code to Python i.e., `pytagi.manual_seed(42)` and the CPP runner. For CPP, taking an example of mnist classification `test/fnn/test_fnn_mnist_cpu.cpp`, we just need to set a seed at the beginning of the script. As you can see, both Python and CPP behave the same way. If we dont manually set the seed, it will randomly be assigned a seed `std::random_device{}()` 
```cpp
void manual_seed(int seed) {
    SeedManager::get_instance().set_seed(seed);
}
```

lhnguyen102 commented on src/bindings/utils_bindings.cpp:
```diff
@@ -169,4 +169,11 @@ void bind_utils(pybind11::module_ &m) {
\ No newline at end of file
+}
+
+void bind_seed_manager(pybind11::module_ &m) {
```
I dont think we need to bind the `SeedManager` class. See my general comment in PR

lhnguyen102 commented on pytagi/__init__.py:
```diff
@@ -1,3 +1,11 @@
+# Create a single instance of SeedManager
+_seed_manager = SeedManager()
+
+def manual_seed(seed: int) -> None:
```
We should do this in the CPP side, so both shared the same function

lhnguyen102 commented on pytagi/tagi_utils.py:
```diff
@@ -299,3 +299,17 @@ def compute_max_min(data: np.ndarray) -> Tuple[np.ndarray, np.ndarray]:
 
         return (np.nanmax(data, axis=0), np.nanmin(data, axis=0))
+
+class SeedManager:
```
This won't be needed if we decide to implement manual_seed in CPP side

lhnguyen102 commented:
@miquelflorensa Could you please add the following code for cuda detection in this PR, I dont want to create a PR for it, I would put it in `src/common.cpp` and create a binding code for Python as well. Thank you!
```cpp
bool is_cuda_available() {
#ifdef USE_CUDA
    return true;
#else
    return false;
#endif
}
```

miquelflorensa commented:
@lhnguyen102 I added the `manual_seed` and `is_cuda_available` functions on `common.cpp` and added their bindings on utils bindings (don't know if we should create another bindings file for them). On the other hand I added a `reset_seed` function that reinitialize the randomizer so when the device is switch to cuda it creates the same parameters as in c++ (I know the way I did it is not the cleanest, I will think on a better solution tomorrow).

lhnguyen102 commented:
@miquelflorensa it is nice to have a `reset_seed()`.I take a look at the CUDA code. The problem with CUDA layer is that it will initialize once again weights and biases when `model.to_device("cuda")` e.g., [linear layer](https://github.com/lhnguyen102/cuTAGI/blob/main/src/linear_layer_cuda.cu#L1708). As you insist on having a clean manual seed setting for both CPP and CUDA, I think you can transfer weights and biases from CPP layer directly to CUDA layer. Here are my suggestions
- Implement a copying function in `BaseLayerCuda` class that copies all weights and biases to CUDA layer and sending it to device
- Call the previous function [here](https://github.com/lhnguyen102/cuTAGI/blob/main/src/linear_layer.cpp#L635)
- Remove the initializing function for weights and biases in [here](https://github.com/lhnguyen102/cuTAGI/blob/main/src/linear_layer_cuda.cu#L1708)
I am OK with the currently implementation that you did. However, I never say no to a better solution. Feel free to get creative on the solution because I only provide the my view based on my exp on cuTAGI!

miquelflorensa commented:
@lhnguyen102 I made a `copy_params_from` function that is called from `to_cuda` functions in Linear and LSTM as you said.

lhnguyen102 commented:
@miquelflorensa It looks perfect to me. Any reason that we cannot apply the same technique to `conv2d` and `tconv2d`?

lhnguyen102 commented on include/bindings/utils_bindings.h:
```diff
@@ -24,7 +24,7 @@ class Utils {
    public:
     Utils();
     ~Utils();
-    std::string get_name() { return "fuck"; }
```
:)

miquelflorensa commented:
@lhnguyen102 I didn’t implement it for the Conv2d and TConv2d layers because parameter initialization is currently handled during the first forward pass. However, I’m now considering whether it might be easier to move all parameter initialization into the forward function. Or should we move them to the constructor of the layer instead?

lhnguyen102 commented:
@miquel-florensa you can remove [this section](https://github.com/lhnguyen102/cuTAGI/blob/main/src/convtranspose2d_layer_cuda.cu#L210) and use the copy function for initializing the parameters. The only thing now we need to take care is the `this->allocate_param_delta()`. This function can be found [here](https://github.com/lhnguyen102/cuTAGI/blob/main/src/base_layer_cuda.cu#L143). A solution is to set a condition to trigger this allocation in the forward pass
```cpp
if (this->training && this->delta_mu_w.size() == 0) {
    this->allocate_param_delta();
}
```

miquelflorensa commented:
@lhnguyen102 I’m still unsure why we need to use the copy functions in Conv and TransposeConv. Since both layers initialize their parameters during the forward pass, they are already on the desired device (CUDA or CPU). I tested the current setup, and the initialization works equally on both devices without issues. Also, I rebased my branch with the latest main version, but now all your changes are appearing in this PR as well. Not sure why that’s happening.

lhnguyen102 commented:
@miquel-florensa You're rights! I was overthinking.

lhnguyen102 commented:
> Also, I rebased my branch with the latest main version, but now all your changes are appearing in this PR as well. Not sure why that’s happening.

This is weird and it marked that your branch is out-of-date with the base branch.

miquelflorensa commented:
@lhnguyen102 Ok, now it's good.

lhnguyen102 commented:
@miquelflorensa how did you do it? :)

miquelflorensa commented:
@lhnguyen102 I went back to the previous commit and I rebased from this same page in GitHub, where there was a button to rebase :)
